src/analysis/asymptotics/superpolynomial_decay.lean

```diff
@@ -13,296 +13,288 @@ import data.polynomial.eval
+  superpolynomial_decay l k (λ x, (p.eval $ k x) * f x) :=
+polynomial.induction_on' p (λ p q hp hq, by simpa [add_mul] using hp.add hq)
+  (λ n c, by simpa [mul_assoc] using (hf.parameter_pow_mul n).const_mul c)
 
-lemma superpolynomial_decay.mono (hf : superpolynomial_decay f)
-  (h : ∀ n, ∥g n∥ ≤ ∥f n∥) : superpolynomial_decay g :=
-(is_O_of_le at_top h).trans_superpolynomial_decay hf
+lemma superpolynomial_decay.mul_polynomial [has_continuous_add β] [has_continuous_mul β]
+  (hf : superpolynomial_decay l k f) (p : polynomial β) :
+  superpolynomial_decay l k (λ x, f x * (p.eval $ k x)) :=
+(hf.polynomial_mul p).congr (λ _, mul_comm _ _)
 
-lemma superpolynomial_decay.eventually_mono (hf : superpolynomial_decay f)
-  (h : ∀ᶠ n in at_top, ∥g n∥ ≤ ∥f n∥) : superpolynomial_decay g :=
-(is_O_iff.2 ⟨1, by simpa only [one_mul] using h⟩).trans_superpolynomial_decay hf
+end comm_semiring
 
-@[simp]
-lemma superpolynomial_decay_zero : superpolynomial_decay (0 : α → 𝕜) :=
-λ c, is_O_zero _ _
+section ordered_comm_semiring
 
-@[simp]
-lemma superpolynomial_decay_zero' : superpolynomial_decay (λ (x : α), (0 : 𝕜)) :=
-superpolynomial_decay_zero
+variables [topological_space β] [ordered_comm_semiring β] [order_topology β]
 
-lemma superpolynomial_decay.add (hf : superpolynomial_decay f) (hg : superpolynomial_decay g) :
-  superpolynomial_decay (f + g) :=
-λ c, is_O.add (hf c) (hg c)
+lemma superpolynomial_decay.trans_eventually_le (hk : 0 ≤ᶠ[l] k)
+  (hg : superpolynomial_decay l k g) (hg' : superpolynomial_decay l k g')
+  (hfg : g ≤ᶠ[l] f) (hfg' : f ≤ᶠ[l] g') : superpolynomial_decay l k f :=
+λ z, tendsto_of_tendsto_of_tendsto_of_le_of_le' (hg z) (hg' z)
+  (hfg.mp (hk.mono $ λ x hx hx', mul_le_mul_of_nonneg_left hx' (pow_nonneg hx z)))
+  (hfg'.mp (hk.mono $ λ x hx hx', mul_le_mul_of_nonneg_left hx' (pow_nonneg hx z)))
 
-lemma superpolynomial_decay.const_mul (hf : superpolynomial_decay f) (c : 𝕜) :
-  superpolynomial_decay (λ n, c * f n) :=
-(is_O_const_mul_self c f at_top).trans_superpolynomial_decay hf
+end ordered_comm_semiring
 
-lemma superpolynomial_decay.mul_const (hf : superpolynomial_decay f) (c : 𝕜) :
-  superpolynomial_decay (λ n, f n * c) :=
-by simpa [mul_comm _ c] using superpolynomial_decay.const_mul hf c
+section linear_ordered_comm_ring
 
-lemma superpolynomial_decay_const_mul_iff_of_ne_zero {c : 𝕜} (hc : c ≠ 0) :
-  superpolynomial_decay (λ n, c * f n) ↔ superpolynomial_decay f :=
-⟨λ h, (is_O_self_const_mul c hc f at_top).trans_superpolynomial_decay h, λ h, h.const_mul c ⟩
+variables [topological_space β] [linear_ordered_comm_ring β] [order_topology β]
 
-lemma superpolynomial_decay_mul_const_iff_of_ne_zero {c : 𝕜} (hc : c ≠ 0) :
-  superpolynomial_decay (λ n, f n * c) ↔ superpolynomial_decay f :=
-by simpa [mul_comm _ c] using superpolynomial_decay_const_mul_iff_of_ne_zero hc
+lemma superpolynomial_decay_iff_abs_tendsto_zero :
+  superpolynomial_decay l k f ↔ ∀ (n : ℕ), tendsto (λ (a : α), |(k a) ^ n * f a|) l (𝓝 0) :=
```

Comment:
Is the RHS the same as `superpolynomial_decay l (λ a, |k a|) (λ a, |f a|)`? If so, that seems like a useful statement.

Reply:
Yes, and a similar thing also applies to `superpolynomial_decay_iff_norm_tendsto_zero` so I've also added the corresponding lemma for that as well.